### ISSUE TITLE
Tests

### ISSUE THREAD

**Issue by KateKate** [redux-effects/redux-effects-fetch#7]
A lot of thanks for your library. Have you receipt how to write integration tests for that in browser? Tried to use https://www.npmjs.com/package/fetch-mock, but it doesn't work. Because it requires 'fetch should be added as a global'. Thanks in advance!

**Comment by ashaffer** [comment]
Ah, tests are an area where redux-effects should excel.  Just replace your middleware :).  E.g.

```javascript
const fetchLog = []

function mockFetch (api) {
  return next => action =>
    action.type === 'EFFECT_FETCH' ? fetchLog.push(action) : next(action)
}

test('test some things', t => {
  render(<SomeThingThatFetches />)
  t.equal(fetchLog.length, 1)
})
```

Does that help or is that not what you were looking for?

**Comment by KateKate** [comment]
yep, less is more ) Thanks!

**Comment by ashaffer** [comment]
Great! np.

**Comment by nikolasleblanc** [comment]
Can you provide an example that deals with mocked responses?

**Comment by dyong0** [comment]
@nikolasleblanc I think you can do that using [redux-effects-fetch-fixture](https://github.com/team-boris/redux-effects-fetch-fixture).

**Comment by nikolasleblanc** [comment]
Trying that now with a mock store and it's almost there. Just can't fail a
test without zone js getting upset about an unhanded promise rejection.
Getting this with chai.expect failures, chai.assert failures, and simple
throws. If I figure it out I'll share.

**Comment by nikolasleblanc** [comment]
Here's a mock store factory, it allows you to provide fixtures, middleware, expected actions (in our case, probably an array of one, but just in case), initial state, and a done call back - note, this is using angular 2's testing suite so done has a fail method. Easy enough to switch out if your done just takes a parameter directly. `done(err)` vs `done.fail(err)`.

The idea here is that you use https://github.com/team-boris/redux-effects-fetch-fixture in your middleware stack instead of https://github.com/redux-effects/redux-effects-fetch, and provide it with fixtures mocking out your api responses.

Also, using https://github.com/flitbit/diff to better inform test failures.

```
import {applyMiddleware} from 'redux';

const deepDiff = require('deep-diff');

export const createMockStore = (
  fixtures = {},
  middlewares = new Array()
) => {
  const mockStore = ({getState, dispatch}) => {
    function createStore() {
      return {
        getState,
        dispatch
      };
    }

    const finalCreateStore = applyMiddleware(
      ...middlewares
    )(createStore);

    return finalCreateStore();
  };
  const configureMockStore = (initialState = {}, expectedActions, done) => {
    return mockStore({
      getState: () => initialState,
      dispatch: (action) => {
        const expectedAction = expectedActions.shift();
        try {
          chai.expect(action).to.deep.equal(expectedAction);
          done();
        } catch (err) {
          done.fail(`
          ${err}
          Deep diff: ${JSON.stringify(deepDiff(action, expectedAction))}
        `);
        }
      }
    });
  };
  return configureMockStore;
};
```

And here's a sample test using a mock store returned by the function above:

```
import {
  it,
  describe
} from '@angular/core/testing';

const effects = require('redux-effects');
const multi = require('redux-multi');
const fetchFixture = require('redux-effects-fetch-fixture');

const responses = require('redux-effects-fetch-fixture').responses;

import * as ManifestConstants from '../constants/manifest-constants';
import * as ManifestActions from '../actions/manifest-actions';

import {
  MAPPED_POPULATED_MANIFEST_BY_ID,
  MANIFEST_GET_BY_ID
} from '../fixtures/manifest-fixture';
import {TCP_GET_BY_ID} from '../fixtures/tcp-fixture';

import {createMockStore} from '../util/mock-store-fixture';

describe('Manifest actions test.', () => {

  describe('aynschronous action creators', () => {
    it('should get a manifest by id', (done) => {

      // 1. Setup expected actions
      const expectedActions = [
        {
          type: 'GET_MANIFEST',
          selected: MAPPED_POPULATED_MANIFEST_BY_ID
        }
      ];

      // 2. Setup fixtures (mocks)
      const fixtures = {
        '//127.0.0.1:8080/services/manifest/123': {
          // this delegates to another fixture
          'GET': () => responses.ok(
            MANIFEST_GET_BY_ID
          )
        },
        '//127.0.0.1:8080/services/tcp/654': {
          // this delegates to another fixture
          'GET': () => responses.ok(
            TCP_GET_BY_ID
          )
        }
      };

      // 3. Setup middleware stack required for testing
      const middlewares = new Array();
      middlewares.push(effects.default);
      middlewares.push(fetchFixture.default(fixtures));
      middlewares.push(multi.default);

      // 4. Create mockstore with empty initial state
      const store = createMockStore(fixtures, middlewares)({}, expectedActions, done);

      // 5. Dispatch action, tests are evaluated in mock
      store.dispatch(ManifestActions.getByIdViaMiddleware(123));
    });
  });
});
```

Question for you testers though, what's your opinion on this? It's certainly handy to be able to test with mock data, but is it _necessary_? If I trust the redux-effects-fetch middleware, do I need to test it's actual behaviour, or can I just test that the initial action dispatched is the EFFECTS_COMPOSE action that I expect, as opposed to the reduced action that the middleware will eventually send to my store?

In other words, if we're doing this declaratively, can I test this declaratively?

**Comment by nikolasleblanc** [comment]
Here's an example of simply testing the action that will be picked up by the middleware:

```
it('should return an EFFECTS_COMPOSE action', () => {
  // test declaratively, do we need to care about how this action is processed by middleware?
  const action = ManifestActions.getByIdViaMiddleware(123);
  chai.expect(action.length).to.equal(1);
  chai.expect(action[0].type).to.equal('EFFECT_COMPOSE');
  chai.expect(action[0].payload.type).to.equal('EFFECT_FETCH');
  chai.expect(action[0].payload.payload.url).to.equal('http://127.0.0.1:8080/services/manifest/123');
  chai.expect(action[0].payload.payload.params.method).to.equal('GET');
  chai.expect(action[0].meta.steps.length).to.equal(1);
  chai.expect(action[0].meta.steps[0].length).to.equal(2);
});
```
